packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -646,22 +646,32 @@ export default createRule<Options, MessageId>({
           .getCallSignaturesOfType(
             getConstrainedTypeAtLocation(services, callback),
           )
-          .map(sig => sig.getReturnType());
+          .map(sig => sig.getReturnType())
+          .map(t => {
+            if (tsutils.isTypeParameter(t)) {
+              return checker.getBaseConstraintOfType(t);
+            }
+
+            return t;
```

Comment:
Optimally, we could use `getConstrainedTypeAtLocation` here once https://github.com/typescript-eslint/typescript-eslint/issues/10438 is tackled.

Reply:
👍 agreed, that'd be nice. Suggestion: I'd put a TODO comment here so it's not lost.